src/later.cpp

```diff
@@ -149,16 +161,25 @@ bool execCallbacks(double timeoutSecs, bool runAll, int loop) {
   Rcpp::RNGScope rngscope;
   ProtectCallbacks pcscope;
 
-  if (!getCallbackRegistry(loop)->wait(timeoutSecs)) {
+  boost::shared_ptr<CallbackRegistry> callback_registry;
+  {
```

Comment:
For some reason, when this block of code was written as:

```c++
    callbackRegistriesMutex.lock();
    callback_registry = getCallbackRegistry(loop);
    callbackRegistriesMutex.unlock();
```

... some of the tests would fail when running tests via RStudio's Test Package command (though not usually when running `devtools::test()` from the R console).

Specifically, these tests:
  https://github.com/r-lib/later/blob/55f63f3/tests/testthat/test-run_now.R#L61-L72

The `run_now(3)` would always run for 3 seconds, and the `result` was `FALSE`. I don't understand what the functional difference is from the current version of the code, so I'm puzzled why it happened.

Reply:
Mystery solved, with @jcheng5. In some other tests which were run earlier, `getCallbackRegistry()` throws an exception, and the mutex never gets released, so the background thread in this test never is able to acquire the mutex and call `getCallbackRegistry()`.

Reply:
Is it still possible for an exception to cause a mutex to never be released, now that you're wrapping invoke() so carefully?